crates/event-streaming/proto/walrus/event/v1alpha/service.proto

```diff
@@ -0,0 +1,164 @@
+
+    optional uint64 checkpoint_sequence_number = 3;
+
+    // The cryptographic proof that ties the events to the checkpoint summary.
+    // This is designed to be upgradeable.
+    optional Proof proof = 4;
+}
+
+message StreamHead {
+    // The Object ID of the dedicated stream head object.
+    optional string object_id = 1;
+}
+
+message PackageSubscription {
+    // The package object ID to subscribe to.
+    optional string package_id = 1;
+}
+
+// ============== Proof Structures ============================
+
+// Proof is an enum that allows the server to send different types of proofs.
+// This enables a future protocol upgrade from the simple V1 proof to the
+// more efficient V2 Merkle proof without breaking the API contract.
+message Proof {
+    oneof proof_type {
+        // V1 Proof: The entire raw contents of the checkpoint.
+        // The client verifies this by hashing the bcs and comparing
+        // it to the digest in the CheckpointSummary. And comparing the
+        // events against the content. Inefficient but simple.
+        sui.rpc.v2beta.CheckpointContents checkpoint_contents = 1;
+
+        // V2 Proof: A compact Merkle proof.
+        MerkleInclusionProof merkle_proof = 2;
+
+        // No-Proof Option: Sent when the client trusts the server and opts out
+        // of verification to save bandwidth and CPU.
+        google.protobuf.Empty proof_none = 3;
+    }
+}
+
+// Represents a Merkle inclusion proof for a checkpoint artifact.
+message MerkleInclusionProof {
+    // The index of the leaf in the Merkle tree.
+    optional uint64 leaf_index = 1;
+
+    // The raw bytes of the event artifact that was hashed to create the leaf.
+    optional bytes leaf_artifact = 2;
+
+    // The list of sibling hashes needed to reconstruct the path to the root.
+    repeated bytes sibling_hashes = 3;
+}
+
+// ============== Non-Inclusion Proof RPC Messages ==============
+
+message GetNonInclusionProofRequest {
+    // The starting checkpoint sequence number for the range to check (inclusive).
+    optional uint64 start_checkpoint_sequence_number = 1;
+
+    // The ending checkpoint sequence number for the range to check (inclusive).
+    optional uint64 end_checkpoint_sequence_number = 2;
+
+    // The selector for the event stream to prove non-inclusion against.
+    oneof stream_selector {
+        PackageSubscription package_id = 3;
+        StreamHead stream_head = 4;
+    }
+}
+
+// Represents a proof that a specific event was not included in a checkpoint.
+// This is only possible with a V2 (Merkle tree) proofing system.
+message NonInclusionProof {
+    // The sequence number of the checkpoint this proof applies to.
+    optional uint64 checkpoint_sequence_number = 1;
+
+    // A valid non-inclusion proof consists of proving the existence of the
+    // two artifacts that are the immediate neighbors of the item being disproven
+    // in the sorted Merkle tree.
+    repeated MerkleInclusionProof adjacent_proofs = 2;
```

Comment:
Is it better to instead have two fields that are optional? One for the left leaf and another for the right leaf.

Reply:
I like the two optional fields better, good idea!